File: Library.Application/Books/Dtos/SearchBooksFilterDto.cs
namespace Library.Application.Books.Dtos
{
	public record SearchBooksFilterDto(
		string? Term,
		Language? Language,
		CoverType? CoverType,
		List<int>? GenreIds,
		bool? AvailableOnly
	)
	{
		public Expression<Func<Book, bool>> Predicate()
		{
            return b =>
			(string.IsNullOrEmpty(Term) || b.Title.ToLower().Contains(Term.ToLower()) || b.Author.Name.ToLower().Contains(Term.ToLower()) || b.ISBN.Value.Contains(Term.ToLower()))
			&& (Language == null || b.Language == Language)
			&& (CoverType == null || b.CoverType == CoverType)
			&& (GenreIds == null || !GenreIds.Any() || b.Genres.Any(g => GenreIds.Contains(g.Id)))
			&& (AvailableOnly != true || b.AvailableCopies > 0);
		}
	}
}

File: library-app-client/src/app/features/book/dtos/search-books-filter.dto.ts
import { CoverType } from "../enums/cover-type.enum";
import { Language } from "../enums/language.enum";

export interface SearchBooksFilterDto {
  term?: string | null;
  language?: Language | null;
  coverType?: CoverType | null;
  genreIds?: number[] | null;
  availableOnly?: boolean | null;
}